CompileVisitor using the visitor pattern and CompiledInjector

koriym opened the pull request:
Moved from #112 

Closes #112 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Ray.Compiler Release Notes

### New Features
- Introduced `CompiledInjector` for enhanced dependency injection performance.
- Added `Compiler` class for script compilation.
- Implemented new script generation and management functions.
- Added support for advanced dependency injection scenarios.
- Introduced `VisitorCompilerTest` for comprehensive testing of `CompileVisitor`.
- Added `FakeModule` for testing purposes.
- New documentation for `ScriptDirNotReadable` and `Unbound` errors added.

### Improvements
- Streamlined dependency injection compilation process.
- Enhanced type safety and error handling.
- Improved performance through pre-compiled scripts.
- Simplified module configuration and binding.
- Consolidated injector setup in tests for better code reuse.

### Breaking Changes
- Deprecated `ScriptInjector` in favor of `CompiledInjector`.
- Removed several deprecated classes and methods, including `DependencyCompilerTest` and `DiCompilerTest`.
- Updated module and injector initialization patterns.
- Changed expected injector types in tests to `CompiledInjector`.

### Deprecations
- Marked multiple classes as deprecated, including `DiCompiler`, `ScriptInjector`, and `CompileInjector`.
- Recommended using `CompilerModule` and `CompiledInjector`.

### Bug Fixes
- Improved error handling for script directory and dependency resolution.
- Fixed issues with dependency injection and compilation processes.

### Documentation
- Completely revised README with clearer installation and usage instructions.
- Added performance comparison and best practices.
- Updated documentation to reflect new class names and usage patterns.
- Introduced documentation for `ScriptDirNotReadable` error with troubleshooting tips.
- Added details for the "Unbound Error" and its resolution strategies.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

koriym committed:
Add compiler and related features

Introduce core classes for the compiler including `AirInjector`, `Compiler`, `CompileVisitor`, `InstanceScript`, and `Scripts`. These additions are essential for enabling the compilation and injection mechanisms. Additionally, add exceptions for specific error handling scenarios and update `composer.json` to remove the unnecessary php-parser dependency.

koriym committed:
Add type annotations and coding standard comments

Modified `AirInjector`, `DependencyCode`, `Compiler`, and `InstanceScript` to improve code readability and maintain coding standards. Added missing type annotations, aligned annotations to coding standards, and integrated comments for static analysis tools.

koriym committed:
Remove obsolete AOP and DI-related classes

Deleted several unused classes including AopCode, Code, DependencyCode, and others related to AOP and Dependency Injection mechanisms. This cleanup aims to streamline the codebase by removing legacy components that are no longer in use.

koriym committed:
Refactor InjectionPoint and related classes for better DI

Simplify the `InjectionPoint::getQualifier` method by leveraging the `ServiceLocator` for retrieving annotations instead of file-based qualifiers. Enhance `CompileInjector` and `AirInjector` by introducing new bindings and psalm-type imports to increase flexibility and maintainability of dependency injection.

koriym committed:
Improve Fake class

Add new bindings and methods for improved functionality.

Added a new binding for FakeAssistedConsumer in FakeToBindModule. Included VisitorInterface usage in FakeInvalidDependency and added an accept method. Applied Attribute and Qualifier annotations in FakeLoggerInject.

koriym committed:
fixup! Remove obsolete AOP and DI-related classes

koriym committed:
Add LazyModule class for creating a lazy module

This class provides a factory method to generate instances of LazyModuleInterface. It is intended for testing purposes and should not be used in production, as it allows creating a module from a callable variable.

koriym committed:
Refactor test setup with CompileInjector

Replaced ScriptInjector with CompileInjector in the setUp method. Utilized LazyModule for instantiating FakeToBindModule to streamline the testing process.

koriym committed:
Remove obsolete testCompileFailureLog function

The function `testCompileFailureLog` was removed due to redundancy. The file_get_contents usage was also removed since it is no longer needed.

koriym committed:
Refactor ContextBindingTest to use a CompileInjector

The private injector property was introduced and initialized in the setUp method to prevent redundant instantiation. This ensures cleaner code and reduces duplication by reusing the same injector across multiple tests.

koriym committed:
fixup! Add LazyModule class for creating a lazy module

koriym committed:
Refactor MultiBindingTest using CompileInjector

koriym committed:
Implement LazyModuleInterface in LazyModule class

The LazyModule class now explicitly implements the LazyModuleInterface, ensuring it adheres to the expected contract. Constructor and __invoke() method are added to manage and return the AbstractModule instance, enhancing compliance and functionality.

koriym committed:
Refactor to use CompileInjector with LazyModule

Switch from ScriptInjector to CompileInjector in both the null_object test script and InjectorFactory for improved dependency management. Leveraging LazyModule ensures more efficient instantiation and better performance in module loading. This refactoring aligns with modern practices for streamline injection handling.

koriym committed:
Refactor tests to use CompileInjector and enhance coverage

Updated tests to replace ScriptInjector with CompileInjector, ensuring consistency and improved testing strategies. Introduced new tests for various scenarios in VisitorCompilerTest, covering instance compilation, dependency resolution, and contextual provider behavior. Added missing bindings and LazyModule implementations to support the changes and ensure test reliability.

koriym committed:
Remove deprecated test files for Normalizer and PrivateProperty.

These test files are no longer in use and have been deprecated. Cleaning up the codebase by deleting these obsolete tests will help in maintaining clarity and reducing clutter.

koriym committed:
Remove obsolete DiCompilerTest.php file

The DiCompilerTest.php file has been deleted, as it is no longer needed or relevant to the current codebase. Removing this obsolete file helps maintain a clean and manageable test directory. This change should not affect the current functionality or coverage of the testing suite.

koriym committed:
Remove unused imports from DevInjectorContext.php

Unused imports DiCompileModule and FakeCarModule were removed to clean up the code. This enhances readability and reduces potential confusion for developers working with the file. Keeping the codebase clean and free of unnecessary dependencies helps maintain a high standard of code quality.

koriym committed:
Remove scriptDir dependency from InjectionPoint class.

This change simplifies the constructor of the InjectionPoint class by removing the scriptDir parameter, which was deemed unnecessary. Modifications have been made in related files to adjust method calls and annotations accordingly. This refactor enhances code maintainability and reduces complexity by removing redundant elements.

koriym committed:
Soothe SA: Add type annotation for $isProd in InjectorFactory

A type annotation is added for the $isProd variable in InjectorFactory to clarify its expected type. This improves code readability and helps with better static analysis by tools.

koriym committed:
Update Symfony dependencies in composer.lock

Upgraded "symfony/dependency-injection" to v7.1.9, "symfony/deprecation-contracts" to v3.5.1, and "symfony/service-contracts" to v3.5.1. These updates bring the latest improvements and fixes from Symfony, ensuring better stability and performance.

koriym committed:
Fix CS: Add type hint to ContextBindingTest injector property

Introduced a type hint for the private $injector property in the ContextBindingTest class to enhance code clarity and type safety. Additionally, renamed ScriptInjectorTest.php to CompileInjectorExtendedScriptInjectorTest.php to better reflect its purpose and alignment with naming conventions.

koriym committed:
Rename DiCompiler.php to src-deprecated directory

This change moves the DiCompiler.php file into the src-deprecated directory to indicate its deprecated status. A minor formatting change removing an empty line was also made in the file. These updates help organize the codebase by clearly separating deprecated components.

koriym committed:
Refactor test assertion for instance type checks

Update test assertions to use assertInstanceOf for type validation in CompileInjectorExtendedScriptInjectorTest. This change improves readability and consistency in the unit tests by utilizing a more expressive assertion method.

koriym committed:
Update ray/di dependency for compatibility

Change the ray/di development dependency to use the dev version aliased as 2.17.3 for ensuring compatibility with the existing project setup. This update addresses potential issues with dependency resolution in the development environment.

koriym committed:
Refactor temporary directory creation in tests.

This commit introduces a private method, `getTmpDir`, to streamline the creation of temporary directories in test cases. By consolidating directory initialization logic, the code becomes cleaner and reduces repetition. This change enhances maintainability and readability across multiple test functions.

koriym committed:
Add runtime type assertions in VisitorCompilerTest

Include assert checks to verify instances implement AcceptInterface in various test cases. This ensures type correctness during acceptance by CompileVisitor, enhancing test reliability and debugging clarity.

koriym committed:
Ignore PHPStan warning on InjectionPoint initialization

Added a comment to suppress a PHPStan warning when assigning the anonymous function to the $injectionPoint variable. This change ensures cleaner code analysis without altering the underlying functionality.

koriym committed:
Refactor test setup to use a new CompileInjector instance

Update test methods to create a new CompileInjector instance using a temporary directory and LazyModule for each test execution. This change improves test isolation and ensures consistency by avoiding shared state between tests. Add a type hint to the getTmpDir function to specify its return type.

koriym committed:
fixup! Update ray/di dependency for compatibility

koriym committed:
Add ScriptDir annotation to Compiler class

Introduced the ScriptDir annotation to enhance the type safety of the $scriptDir parameter in the Compiler class. This change ensures consistency and correctness throughout the codebase by leveraging psalm-import-type for better type management.

koriym committed:
Add ray/di dependency and update autoload paths

This commit introduces the "ray/di" dependency with version ^2.17 in the composer.json file, ensuring compatibility and additional features. The autoload paths have been updated to include only the "src" directory, removing deprecated source paths for cleaner code structure. Unused branch-alias settings have also been removed to streamline the configuration.

koriym committed:
Update dependencies and deprecate DiCompiler class.

Upgraded phpstan/phpstan to version 2.0.3 for improved static analysis. Removed doctrine/annotations dependency from composer.json as it is no longer required. Marked DiCompiler class as deprecated in favor of CompileInjector for future compatibility.

koriym committed:
Refactor docblock style in Compiler.php

Condense docblock import statement into a single line for clarity. This change maintains functionality while improving code readability and consistency. No functional changes were made to the logic of the Compiler class.

koriym committed:
Remove unused doctrine/annotations dependency

The dependency "doctrine/annotations" is no longer necessary and has been removed from the composer.json file. This change helps to simplify the dependency management and reduce potential maintenance overhead.

koriym committed:
Remove unnecessary Ray\Di components from config

The redundant Ray\Di components were removed from the `composer-require-checker.json` configuration file. This streamlines the dependencies and ensures the configuration focuses on required elements only, reducing maintenance overhead and potential for errors.

koriym committed:
Add PhpParser to composer-require-checker configuration

Note: PhpParser is only used in deprecated-src.

This change updates the configuration file by adding "PhpParser" to the list of globally recognized symbols. This ensures that the composer-require-checker tool accounts for "PhpParser" when verifying dependencies, preventing false positives in projects using this symbol.

koriym committed:
fixup! Add ray/di dependency and update autoload paths

koriym committed:
Fix incorrect assertion in InjectorFactoryTest

The test now correctly asserts that the injector is an instance of CompileInjector instead of ScriptInjector. This change ensures the test accurately reflects the expected behavior of the injector creation process.

koriym committed:
Refactor LazyModule initialization in tests.

Replaced the usage of LazyModule::getInstance with direct instantiation of LazyModule for improved readability and maintainability in test files. Removed the unnecessary getInstance method from the LazyModule class, simplifying the code. This change ensures a more straightforward and consistent approach to handling LazyModule objects across the codebase.

koriym committed:
fixup! fixup! Add ray/di dependency and update autoload paths

koriym committed:
Remove default values for $ip parameter in functions.

The default empty array for the $ip parameter was removed to ensure accurate and explicit input is provided when these functions are called. This change enhances code clarity and enforces stricter argument requirements which help prevent errors due to unintended default values.

koriym committed:
Avoit static functions in AirInjector

static functions does NOT work in multiple application enviroments. There is no problem for normal single application. But when multiple application use each AirInjector with one runtime. static function can't handle different $scritDir.

Extract function initialization logic into a dedicated private method to improve code readability and maintainability. This change reduces redundancy and ensures that functions are only set up once. Additionally, modified the $ip property type to include null for better type safety.

koriym committed:
Add InjectionPoint test

Implemented the FakeOilProvider and FakeOilInterface for dependency injection into the FakeCar class. Added corresponding tests to ensure successful oil injection and integration with the AirInjector and VisitorCompiler modules.

koriym committed:
Remove unused InjectionPoint and update script handling

Removed the 'InjectionPoint' logic and associated dependencies across the codebase, simplifying the handling of dependency instances. Updated function signatures and comments to align with these changes. Introduced constants for string literals to improve code maintainability in 'InstanceScript.php' and deprecated the old constructor for 'InjectionPoint'.

koriym committed:
Enable unused baseline entry detection in Psalm config

Set `findUnusedBaselineEntry` to true to identify obsolete entries in the Psalm baseline file. This update is aimed at keeping the baseline entries relevant and manageable. Additionally, the `findUnusedCode` setting is disabled to prevent false positives during initial checks.

koriym committed:
Fix InjectionPoint namespace in InstanceScript.php

Corrected the namespace for InjectionPoint to ensure proper reference within the codebase. This change resolves any potential issues with class instantiation errors due to incorrect namespace usage.

koriym committed:
Add autoloading for standalone functions in Composer

This commit updates `composer.json` to include standalone function files for autoloading under the "files" key. Two new PHP files, `prototype.php` and `singleton.php`, were added to handle specific functions and improve dependency management. These changes enhance the autoload mechanism by directly including helper functions in the Composer setup.

koriym committed:
Refactor singleton and prototype argument creation

Updated the method to handle file paths in the creation of singleton and prototype arguments, improving code clarity and maintainability. The refactor uses `str_replace` to modify the index for file path creation, and it adjusts the logic to differentiate between singletons and prototypes. Additionally, enhanced the formatting of the instance bindings for better readability.

koriym committed:
Refactor AirInjector: Simplify instance retrieval

Removed redundant methods and assertions related to instance file management for a cleaner and more efficient process. Simplified the `getInstance` method by directly handling file existence checks and exceptions, eliminating unnecessary code. This change improves maintainability and aligns with modern coding practices.

koriym committed:
Refactor CompileInjector to use AirInjector

Replaced internal dependency management logic with AirInjector to streamline and simplify the codebase. This change eliminates redundant functions, clarifies the injection process, and relies on AirInjector for instance retrieval and compilation.

koriym committed:
Update tests with script path and class changes

Modified `CompileInjectorTest` to include a fake module during compile and ensure the bindings log file exists before getting instances. Updated `VisitorCompilerTest` to use script paths for instantiation, ensuring consistency with new script-based instantiation logic. These changes enhance test reliability and alignment with recent codebase updates.

koriym committed:
Add initial Qodana configuration file

Introduce a `qodana.yaml` file to configure Qodana static code analysis. This setup establishes a baseline profile for inspections, outlines specific exclusions, and specifies PHP 7.2 for CI/CD pipeline management.

koriym committed:
fixup! Update tests with script path and class changes

koriym committed:
Refactor and update dependencies for stability.

Added type annotations and exception handling to improve code clarity and reliability. Updated the PHPStan dependency to a newer version for enhanced static analysis and to benefit from bug fixes. These changes help maintain better code quality and alignment with best practices.

koriym committed:
Replace TestClass with FakeTestClass in tests

This commit introduces a new FakeTestClass and replaces instances of TestClass in unit tests with FakeTestClass. Redundant classes and attributes related to TestClass are removed to streamline the test files and prevent namespace conflicts. These changes will improve maintainability and clarity in the test suite.

koriym committed:
Refactor instance creation logic for clarity

Consolidate the logic for adding prototype or singleton instances into a single conditional expression, improving code readability and maintainability. This change eliminates redundant checks and ensures that the corresponding method is invoked directly based on the $isSingleton condition.

koriym commented:
@coderabbitai review

koriym committed:
Add CompileNullObject invocation in CompileInjector

This change ensures that a CompileNullObject instance is called during the CompileInjector process. It enhances the compilation step by handling additional container and script directory logic, improving overall functionality.

koriym committed:
Refactor null object handling in dependency mapping.

Removed unnecessary return statements in the lambda function for cleaner and more consistent code. This change aligns with the intended use and simplifies the logic.

koriym committed:
Add binding for FakeNullInterface in FakeCarModule

This update introduces a new binding for FakeNullInterface with a null implementation in FakeCarModule. It ensures compatibility in tests requiring this interface without actual implementation.

koriym commented on src/Compiler.php:
```diff
@@ -0,0 +1,63 @@
+        // Compile dependencies
+        $compileVisitor = new CompileVisitor($container);
+        $container->map(static function (DependencyInterface $dependency, string $key) use ($scripts, $compileVisitor): DependencyInterface {
+            assert($dependency instanceof AcceptInterface);
+            $script = $dependency->accept($compileVisitor);
+            assert(is_string($script));
```
This is *compiler*. So never exposed in the prod context. Still?

koriym committed:
Update src/Compiler.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

koriym commented on src/CompileInjector.php:
```diff
@@ -34,207 +27,44 @@
-    public function compile(): void
+    public function compile(AbstractModule $module): void
     {
-        $module = (new InstallBuiltinModule())(($this->lazyModule)());
+        $module = (new InstallBuiltinModule())($module);
         (new FilePutContents())(sprintf('%s/_bindings.log', $this->scriptDir), (string) $module);
         (new Bind($module->getContainer(), ''))->annotatedWith(ScriptDir::class)->toInstance($this->scriptDir);
-        (new DiCompiler($module, $this->scriptDir))->compileContainer();
+        (new Bind($module->getContainer(), InjectorInterface::class))->toInstance(new AirInjector($this->scriptDir));
+        (new Compiler())->compile($module, $this->scriptDir);
```
After reviewing this code, I find that the suggestion is not appropriate. Here is why

The point about changing the method signature is incorrect: the signature of the compile() method is already designed to accept an AbstractModule parameter.


The compile() method signature is already designed to accept an AbstractModule parameter.
Looking at the entire code base, this is not a new change, but the original design.
Other test and sample code also uses this format.


The point about the need to update the documentation is also incorrect: the


The use of Compiler and AirInjector in the CompileInjector class is an existing feature.
The README.md already describes this feature and no additional updates are required.

koriym commented on tests/AirInjectorTest.php:
```diff
@@ -0,0 +1,147 @@
+
+    public function estInjectionPoint(): void
+    {
+        // 依存クラスを作成
+        $engineProviderFile = $this->scriptDir . '/FakeEngineProvider.php';
+        file_put_contents($engineProviderFile, '<?php
+       namespace Ray\Compiler;
+       class FakeEngineProvider { 
+           public function __construct($ip) {} 
+       }
+   ');
+        require $engineProviderFile;
+
+        $carFile = $this->scriptDir . '/FakeCar.php';
+        file_put_contents($carFile, '<?php
+       namespace Ray\Compiler;
+       class FakeCar { 
+           public function setEngine(FakeEngineProvider $engine) {} 
+       }
+   ');
+        require $carFile;
+
+        // エンジンプロバイダーのインジェクションコード
+        $providerInstanceFile = $this->scriptDir . '/Ray_Compiler_FakeEngineProvider-' . Name::ANY . '.php';
+        file_put_contents($providerInstanceFile, '<?php
+       $instance = new \Ray\Compiler\FakeEngineProvider($injectionPoint());
+       return $instance;
+   ');
+
+        // 車のインスタンスコード
+        $carInstanceFile = $this->scriptDir . '/Ray_Compiler_FakeCar-' . Name::ANY . '.php';
+        file_put_contents($carInstanceFile, '<?php
+       $instance = new \Ray\Compiler\FakeCar();
+       $instance->setEngine($prototype("Ray_Compiler_FakeEngineProvider-", ["Ray\\Compiler\\FakeCar", "setEngine", "engine"]));
+       return $instance;
+   ');
+
+        $this->expectException(InjectionPointUnbound::class);
+        $this->injector->getInstance(FakeCar::class);
+    }
```
Don't bother 'src-deprecated' sources. Keep it.

koriym replied:
Got it. Delete whole method.

koriym commented on tests/MultiBindingTest.php:
```diff
@@ -34,10 +35,11 @@ class MultiBindingTest extends TestCase
-        $this->injector = new ScriptInjector(__DIR__ . '/tmp', static function () {
-            return new FakeMultiBindingsModule();
-        });
+        @mkdir(__DIR__ . '/tmp/mulit-bindings');
+        $this->injector = new CompileInjector(
+            __DIR__ . '/tmp/mulit-bindings',
+            new LazyModule(new FakeMultiBindingsModule())
+        );
```
Nonted.

koriym committed:
Enable cleanup in AirInjectorTest teardown method

Uncommented cleanup logic in the teardown method to ensure test artifacts are properly removed after each test run. This prevents potential side effects and ensures a cleaner test environment.

koriym commented on tests/AirInjectorTest.php:
```diff
@@ -0,0 +1,147 @@
+    {
+//        array_map('unlink', glob($this->scriptDir . '/*'));
+//        rmdir($this->scriptDir);
+    }
```
Done!

koriym committed:
Update src/CompileInjector.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

koriym committed:
Update tests/AirInjectorTest.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

koriym committed:
Refactor null handling and tidy up dependencies.

Refactored `CompileNullObject` to ensure better conformity with interfaces by introducing `NullDependency`. Added minor code cleanup in `Compiler` and improved test utility functions in `AirInjectorTest` by including additional standard operations like `glob` and `rmdir`. These changes improve readability, maintainability, and adherence to design principles.

koriym committed:
Remove unused test method from AirInjectorTest

The `estInjectionPoint` method was removed as it was unused and redundant. This cleanup improves the clarity and maintainability of the test suite without affecting functionality.

koriym committed:
Remove qodana.yaml configuration file

This file is no longer needed as part of the project. All related Qodana analysis setup and configurations have been removed to simplify the codebase.

koriym commented on qodana.yaml:
```diff
@@ -0,0 +1,44 @@
+      - src/CompileInjector.php
+  - name: PhpPluralMixedCanBeReplacedWithArrayInspection
+    paths:
+      - src/InstanceScript.php
```
Removed.

koriym committed:
Fix misplaced code coverage annotation in Compiler

Moved the `@codeCoverageIgnoreStart` annotation to the correct position to ensure proper exclusion during code coverage analysis. This prevents unintended behavior when generating code coverage reports.

koriym committed:
Remove unused InjectionPointUnbound import from test

The InjectionPointUnbound exception import was redundant and unused in the AirInjectorTest file. Cleaning up unnecessary imports improves code readability and reduces clutter.

koriym committed:
Remove CompileNullObject class and its invocation

The CompileNullObject class and its usage were removed, likely to simplify the code and reduce unnecessary complexity. Its functionality appears to no longer be required in the CompileInjector process.

koriym commented on tests/AirInjectorTest.php:
```diff
@@ -0,0 +1,152 @@
+        $this->assertInstanceOf(InjectorInterface::class, $injector);
+    }
+
+    public function testWithCompiler(): void
+    {
+        $tmpDir = __DIR__ . '/tmp/testWithCompiler';
+        @mkdir($tmpDir);
+        $module = new FakeCarModule();
+        (new Compiler())->compile($module, $tmpDir);
+        $injector = new AirInjector($tmpDir);
+        $instance = $injector->getInstance(FakeCarInterface::class);
+        $this->assertInstanceOf(FakeCar::class, $instance);
+    }
```
Not likely? tearDown method supposed to cleanup job.

koriym commented on src/InjectionPoint.php:
```diff
@@ -6,34 +6,38 @@
-    /** @param ScriptDir $scriptDir */
-    public function __construct(ReflectionParameter $parameter, string $scriptDir)
+    /** @deprecated use getInstance */
+    public function __construct(ReflectionParameter $parameter)
     {
         $this->parameter = $parameter;
-        $this->scriptDir = $scriptDir;
+    }
```
Fixed.

koriym committed:
Mark IpQualifier as deprecated with version and reason.

Updated the @deprecated annotation in IpQualifier to include the version (1.11.0) and an explanation for its deprecation. This improves clarity regarding the class's status and its replacement by the CompileVisitor pattern.

koriym commented on src-deprecated/IpQualifier.php:
```diff
@@ -5,9 +5,9 @@
-
 use function serialize;
 
+/** @deprecated  */
```
Fixed!

koriym commented on src/AirInjector.php:
```diff
@@ -0,0 +1,118 @@
+
+        /** @psalm-suppress  UnsupportedPropertyReferenceUsage */
+        $singletons = &$this->singletons;
+        $scriptDir = $this->scriptDir;
+        /** @var mixed $instance */
+        $instance = require $scriptFile;
+
+        // Save singleton
+        /** @psalm-suppress UndefinedVariable */
+        if (isset($isSingleton) && $isSingleton) {
+            /** @var object $instance */
+            $this->singletons[$dependencyIndex] = $instance;
+        }
+
+        /** @pslam-var T $instance */
+        return $instance;
+    }
+
+    private function registerLoader(): void
+    {
+        if (in_array($this->scriptDir, self::$scriptDirs, true)) {
+            return;
+        }
+
+        if (self::$scriptDirs === []) {
+            spl_autoload_register(
+                static function (string $class): void {
+                    foreach (self::$scriptDirs as $scriptDir) {
+                        $file = sprintf('%s/%s.php', $scriptDir, str_replace('\\', '_', $class));
+                        if (file_exists($file)) {
+                            require_once $file; // @codeCoverageIgnore
+                        }
+                    }
+                }
+            );
+        }
+
+        self::$scriptDirs[] = $this->scriptDir;
+    }
+}
```
Like autoloader, this is a must-have feature.

koriym committed:
Refactor AirInjector with assertions and cleanup

Added an assertion to ensure singletons is an array and removed commented-out unused code for better clarity. Updated comments to provide improved context for variables and streamlined the script loading section.

koriym commented on src/AirInjector.php:
```diff
@@ -0,0 +1,118 @@
+
+        /** @pslam-var T $instance */
+        return $instance;
+    }
```
Done!

koriym committed:
Update src-function/singleton.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

koriym committed:
Update src-function/prototype.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

koriym committed:
Refactor file handling and add custom exception.

Replaced generic RuntimeException with ScriptFileNotFound for missing file scenarios, improving error clarity. Used `realpath` for scriptDir in AirInjector to normalize paths and prevent issues. Adjusted comments and logic for better readability and maintainability.

koriym committed:
Refactor aspect binding handling in pushAspectBind.

Replaced `unserialize` and removed unused `is_iterable`, simplifying the aspect binding logic. Leveraged `getBindings` method for cleaner and more maintainable code.

koriym committed:
Revert "Refactor file handling and add custom exception."

This reverts commit 92dd7baa1e4427012a34999d197618f8207b74f5.

koriym committed:
Use `realpath` to resolve absolute script directory paths.

This change ensures the script directory path is converted to an absolute and canonical form, preventing potential issues with relative paths or symbolic links. It improves code reliability and consistency when referencing paths.

koriym commented on src/InstanceScript.php:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        $this->formerLines[] = sprintf('$instance->bindings = [%s    %s%s];', PHP_EOL, implode(', ' . PHP_EOL . '    ', $interceptors), PHP_EOL);
+    }
```
You are right. Changed to 'aopBind->getBindings()' to retrieve bindings. (can't remember why I do so...)

koriym commented on tests/Fake/FakeInvalidDependency.php:
```diff
@@ -25,4 +26,8 @@ public function setScope($scope)
     {
     }
+
+    public function accept(VisitorInterface $visitor)
+    {
+    }
```
It's OK for FakeClass.

koriym committed:
Add validation for readable script directory in AirInjector

Introduce a `ScriptDirNotReadable` exception to handle cases where the provided script directory is either non-existent or not readable. Updated `AirInjector` to validate the directory on instantiation and added unit tests to ensure correct behavior and exception handling.

koriym commented on src-function/prototype.php:
```diff
@@ -0,0 +1,26 @@
+    if (!$file || !file_exists($file)) {
+        throw new \RuntimeException(sprintf('File not found: %s', $filePath));
+    }
+    if (!str_starts_with($file, realpath($scriptDir))) {
```
Fixed!

koriym commented on src-function/singleton.php:
```diff
@@ -0,0 +1,27 @@
+        if (!$scriptFile || !file_exists($scriptFile)) {
+            throw new \RuntimeException(sprintf('File not found: %s', $filePath));
+        }
+        if (!str_starts_with($scriptFile, realpath($scriptDir))) {
```
Fixed!

koriym committed:
Add type annotations for array bindings in InstanceScript

This change introduces explicit type annotations for array bindings and AOP bindings within InstanceScript. It improves code clarity and helps with static analysis, reducing the potential for type-related errors.

koriym committed:
Refactor file handling and add custom exception.

Replaced generic RuntimeExceptions with a specialized ScriptFileNotFound exception for better error clarity. Simplified file path resolution logic by removing redundant checks and assertions. Enhanced code readability and maintainability by refining structure and comments.

koriym committed:
Add a new file to the repository

This commit introduces a new file with mode 100644. Ensure permissions are appropriate for standard use and align with project conventions.

koriym committed:
Refactor singleton and prototype handling in DI compiler

Updated the dependency injection compiler to streamline the handling of singleton and prototype instances. Introduced a consistent approach for processing $dependencyIndex in both singleton and prototype implementations, ensuring better maintainability and test coverage. Adjusted related tests and modules to align with these changes.

koriym committed:
Remove unused code and dependencies in test files

Refactored `CompileInjectorExtendedScriptInjectorTest` and `AirInjectorTest` by removing unused properties, setup methods, and redundant tests. Simplifies test logic and ensures a cleaner, more maintainable codebase.

koriym committed:
fixup! Refactor singleton and prototype handling in DI compiler

koriym commented on src-function/prototype.php:
```diff
@@ -0,0 +1,23 @@
+
+    // $scriptDir, $Singletons and $ip can be used in the included file
+    return require $file;
+};
```
$scriptDir is sanitized(realpath) in upper layer.

koriym commented on src/InstanceScript.php:
```diff
@@ -0,0 +1,185 @@
+    private $implementsSetContext = false;
+
+    /** @var bool|null */
+    private $isSingleton = null;
```
Done!

koriym commented on tests/CompileInjectorExtendedScriptInjectorTest.php:
```diff
@@ -0,0 +1,368 @@
+    }
+
+    /** @return non-empty-string */
+    private function getTmpDir(string $subDirectory): string
+    {
+        $tmpDir = __DIR__ . '/tmp/' . $subDirectory;
+        @mkdir($tmpDir);
+
+        return $tmpDir;
+    }
```
Ok for test code.

koriym commented on tests/CompileInjectorExtendedScriptInjectorTest.php:
```diff
@@ -0,0 +1,368 @@
+
+        return $tmpDir;
+    }
+}
```
Ok for test.

koriym committed:
Refactor variable usage in AirInjector registerLoader function.

Replaced direct property access with a local variable for better readability and consistency. Removed unnecessary `@codeCoverageIgnore` comment as it is no longer required.

koriym committed:
Remove isSingleton parameter from pushProviderContext

Simplified the pushProviderContext method by removing the unused isSingleton parameter. Adjusted all relevant calls and removed the associated property to streamline the code and reduce redundancy.

koriym commented on src/AirInjector.php:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        self::$scriptDirs[] = $this->scriptDir;
+    }
```
Fixed.

koriym commented on src/AirInjector.php:
```diff
@@ -0,0 +1,123 @@
+                static function (string $class): void {
+                    foreach (self::$scriptDirs as $scriptDir) {
+                        $file = sprintf('%s/%s.php', $scriptDir, str_replace('\\', '_', $class));
+                        if (file_exists($file)) {
+                            require_once $file;
+                        }
```
Security Considerations

Basic path checks with file_exists are sufficient
Additional class name validation is unnecessary as PHP's namespace rules already enforce valid characters
Directory traversal protection isn't critical for internal usage


Performance Aspects

require_once vs require: While require is slightly faster, require_once is safer and the performance difference is negligible in this context
Capturing $this->scriptDir in a local variable is a small but useful optimization


Design Decisions

Keeping the implementation simple is preferred
No need for additional static variables like $isLoaderRegistered
Security review comments suggesting path traversal protection can be safely ignored for this internal use case



The current implementation strikes a good balance between functionality, security, and simplicity, making it suitable for its intended purpose without overengineering.

koriym committed:
fixup! Refactor variable usage in AirInjector registerLoader function.

koriym committed:
Refactor comments in AirInjector for clarity.

Simplified and clarified comments in the getInstance method by removing redundant or unclear notes. Updated comment descriptions to be more precise and directly explain variable usage.

koriym committed:
Replace AirInjector with CompiledInjector

AirInjector and associated test files have been removed, and all references have been replaced with CompiledInjector. This improves consistency and ensures pre-compilation of all bindings without runtime support for unknown classes.

koriym committed:
Enhance documentation and improve Compiler class annotations

Updated README with clearer usage instructions for production and development, emphasizing performance and configuration best practices. Improved `Compiler` class annotations to outline its functionality and steps in the compilation process for better clarity.

koriym committed:
fixup! Replace AirInjector with CompiledInjector

koriym commented on src/CompileInjector.php:
```diff
@@ -39,202 +32,39 @@
-    public function compile(): void
+    public function compile(AbstractModule $module): void
     {
-        $module = (new InstallBuiltinModule())(($this->lazyModule)());
+        $module = (new InstallBuiltinModule())($module);
         (new FilePutContents())(sprintf('%s/_bindings.log', $this->scriptDir), (string) $module);
         (new Bind($module->getContainer(), ''))->annotatedWith(ScriptDir::class)->toInstance($this->scriptDir);
-        (new DiCompiler($module, $this->scriptDir))->compileContainer();
+        (new Bind($module->getContainer(), InjectorInterface::class))->toInstance(new CompiledInjector($this->scriptDir));
+        (new Compiler())->compile($module, $this->scriptDir);
```
This evaluation misses the point. The original issue "Handling permission and disk space issues" does not require refactoring into multiple methods. We only need to:

Remove silent failures (@)
Add proper error messages
Handle disk/permission errors

The code structure itself is fine, and splitting it would add unnecessary complexity. The focus should be on error handling, not architectural changes.

koriym replied:
This suggestion misses the mark. The current implementation already handles file operation errors appropriately:

"Throwing specific exceptions for directory creation failures" - The code already throws FileNotWritable with appropriate error contexts.
"Handling permission and disk space issues" - These are handled through the chain of operations (mkdir -> tempnam -> file_put_contents -> rename), where failures throw FileNotWritable.
"Removing the @ operator" - The @ operator is intentionally used only for unlink() of temporary files where errors are expected and irrelevant. Other critical operations are properly checked.
"Adding specific error messages" - The generic FileNotWritable exception with filename is sufficient since the error context is clear from the operation sequence.

The implementation is concise yet robust, with appropriate error handling for filesystem operations. The suggestion would add unnecessary complexity without meaningful benefits.

koriym committed:
[ci skip] Add exception handling for unwritable binding log files

Introduce a `FileNotWritable` exception in `compile()` to handle cases where the binding log file cannot be written. This ensures more robust error handling and prevents silent failures during compilation.

koriym committed:
Refactor functions and update code style configuration

Added strict types declarations and improved formatting in `singleton` and `prototype` functions for better readability and consistency. Updated `composer.json` and `phpcs.xml` to include the `src-function` directory in code style checks and fixes.

koriym committed:
Fix docblock punctuation in CompileInjector.php

Added a missing period to the @throws annotation for consistency and clarity. This change improves code documentation without altering functionality.

koriym committed:
Add detailed PHPDoc comments to singleton and prototype

Improved code documentation by adding descriptive PHPDoc blocks for the `singleton` and `prototype` functions. These include parameter details, return types, and potential exceptions, enhancing clarity and maintainability.

koriym committed:
Add debug utility and comment in singleton function

Introduced the `var_dump` function for debugging purposes and added a clarifying comment in the `singleton` function. These updates aim to enhance code understanding and provide tools for debugging.

koriym committed:
Set scriptDir to its real path in CompiledInjector

Ensure the provided script directory is stored as its absolute path using `realpath()`. This improves consistency and prevents issues with relative paths in further operations.

koriym committed:
Update return type doc for dependency resolution function

Corrected the return type in the documentation from "mixed" to "object" for clarity. This ensures the doc accurately represents the returned instance type. No functional changes have been made to the code.

koriym committed:
Refactor ScriptDir handling and update Injector binding

Annotated the `$scriptDir` parameter with `#[ScriptDir]` for stricter typing and introduced the `ScriptDir` annotation. Updated the `InjectorInterface` binding to use `CompiledInjector::class` instead of creating an instance directly, improving dependency management.

koriym committed:
Refactor module installation into Compiler class

Moved built-in module installation logic from CompileInjector to Compiler. This improves separation of concerns by centralizing module-specific operations in the Compiler class and simplifies CompileInjector. The changes enhance maintainability and code organization.

koriym committed:
Add CompilerModule and improve dependency handling

Introduced `CompilerModule` for managing compiler-related bindings and streamlined its installation in `Compiler`. Enhanced handling of script directories in `CompiledInjector` and updated the prototype function and tests to support singletons. These changes improve modularity, readability, and maintainability of the compiler code.

koriym committed:
Move modules to `src-deprecated` and mark them as deprecated.

Modules `InstallBuiltinModule`, `MapModule`, `PramReaderModule`, and `DiCompileModule` were moved to `src-deprecated` and annotated with deprecation comments. These are now superseded by `CompilerModule`.

koriym committed:
fixup! Remove CompileNullObject class and its invocation

koriym committed:
Revert `CompileInjector` to previous released version to keep compatibility

Replaced `CompiledInjector` with `CompileInjector` in tests and refactored the `CompileInjector` class for better modularization and to enhance support for pre-defined dependency resolution. Updated methods to improve singleton management, loader registration, and script compilation handling.

koriym committed:
Deprecate `CompileInjector` and add replacement utilities

Mark `CompileInjector` as deprecated and move it to the `src-deprecated` folder. Introduce `CompileNullObject` and `NullDependency` classes as part of the migration path. These changes support transitioning to `Compiler` and `CompiledInjector` while retaining backward compatibility.

koriym committed:
fixup! Deprecate `CompileInjector` and add replacement utilities

koriym committed:
Update README with detailed production and development usage

Expanded the README with comprehensive instructions for using Ray.Compiler in both production and development environments. Added sections on installation, Docker integration, Composer script integration, and best practices. Improved clarity around pre-compilation, dependency injection, and performance advantages.

koriym committed:
Expand autoload paths to include deprecated source directory.

Added 'src-deprecated' directory to the PSR-4 autoload configuration in `composer.json`. This ensures compatibility with deprecated code while maintaining proper autoloading.

koriym commented on src-deprecated/MapModule.php:
```diff
@@ -8,6 +8,7 @@
 use Ray\Di\MultiBinding\Map;
 use Ray\Di\MultiBinding\MapProvider;
 
+/** @deprecated Use CompilerModule */ */
```
Fixed!

koriym commented on src-deprecated/PramReaderModule.php:
```diff
@@ -8,6 +8,7 @@
 use Koriym\ParamReader\ParamReaderInterface;
 use Ray\Di\AbstractModule;
 
+/** @deprecated Use CompilerModule */ */
```
Fixed!

koriym commented on src-deprecated/DiCompileModule.php:
```diff
@@ -0,0 +1,30 @@
+use Ray\Compiler\Annotation\Compile;
+use Ray\Di\AbstractModule;
+
+/** @deprecated Use CompilerModule */ */
```
Fixed!

koriym commented on src-deprecated/InstallBuiltinModule.php:
```diff
@@ -7,9 +7,9 @@
-
 use function count;
 
+/** @deprecated Use CompilerModule */ */
```
Fixed!

koriym committed:
Swap parameter order in Compiler::compile method

Reordered parameters in the `compile` method to place `$scriptDir` before `$module`. Updated related test cases to reflect the new method signature for consistency and correctness.

koriym committed:
Refactor tests to use `CompiledInjector` and `Compiler`.

Replaced the now-deprecated `CompileInjector` with `CompiledInjector` and updated test setups to use the new `Compiler` class for script generation. Removed `LazyModuleInterface` in favor of direct `AbstractModule` implementations where applicable. Adjusted all related test files and their assertions for consistency with the new approach.

koriym committed:
Update injector implementation to use CompiledInjector

Replaced occurrences of CompileInjector with the updated CompiledInjector across the codebase. Refactored related methods and tests to align with the new implementation, ensuring compatibility and proper functionality. This change reflects the updated architecture of the injector system.

koriym committed:
Update PHPUnit dependency version in composer.json

Reordered and updated the PHPUnit version to include ^8.5.41. This ensures compatibility with newer updates while maintaining support for existing configurations. No functional changes introduced.

koriym committed:
Remove unused multi-binding logic from CompilerModule

The check for multi-bindings and related bindings to MapProvider were removed as they are no longer needed. This simplifies the module and reduces unnecessary code. Functionality related to script directory and injector bindings remains unchanged.

koriym committed:
Ensure `nikic/php-parser` is required in CompileInjector

Add a check for the existence of the `PhpParser\ParserFactory` class in the `CompileInjector` constructor. This prevents misconfiguration by ensuring the required dependency is available, throwing a `LogicException` if it is missing.

koriym committed:
Refactor CompilerModule imports to remove unused dependencies.

Removed unnecessary `count` function import and unused `Map` and `MapProvider` classes. This streamlines the code and eliminates redundant dependencies, improving maintainability.

koriym committed:
Refactor type definitions into a dedicated `Types` class.

Moved psalm type definitions from `CompileInjector` to a new `Types` class for improved modularity and maintainability. Updated all relevant references to use the new `Types` class.

koriym committed:
Add new dependencies for annotations and require-checker

Included `doctrine/annotations` to handle annotation parsing and `maglnet/composer-require-checker` for development dependency checks. This update ensures better compatibility and tooling support for the project.

koriym committed:
Update composer-require-checker excludes with PhpParser types

Expanded the list of excluded symbols to include specific PhpParser-related types and classes. This ensures proper compatibility and avoids unnecessary dependency checks for these symbols in the project.

koriym committed:
Define type annotations and add constructor coverage ignore

Type definitions for Ray.Compiler have been documented using psalm annotations. Additionally, the private constructor is now marked with @codeCoverageIgnore to exclude it from test coverage analysis.

koriym committed:
Remove redundant injector interface condition check

The condition for the `RAY_DI_INJECTOR_INTERFACE` was unnecessary and has been removed to simplify the code. This improves maintainability and ensures that only relevant interface checks are retained.

koriym committed:
Change implemented interface to ScriptInjectorInterface

Updated the CompiledInjector class to implement ScriptInjectorInterface instead of InjectorInterface. This ensures compatibility with the expected interface and aligns the class usage with its intended purpose.

CachedInjectorで保存される

koriym committed:
Rename LazyModule.php and mark it as deprecated

The LazyModule.php file was moved to the src-deprecated directory and marked as deprecated. Users are advised to use Compiler and CompiledInjector as replacements moving forward. This change aids in phasing out older code in favor of newer alternatives.

koriym committed:
Add test for ScriptDirNotReadable exception in CompiledInjector

This commit introduces a new test to ensure that the CompiledInjector throws a ScriptDirNotReadable exception when provided with an unreadable or non-existent script directory. This improves test coverage and ensures proper error handling behavior.

koriym committed:
fixup! Add new dependencies for annotations and require-checker

koriym committed:
fixup! Add test for ScriptDirNotReadable exception in CompiledInjector

koriym committed:
fixup! Change implemented interface to ScriptInjectorInterface

koriym commented on src-deprecated/NullDependendy.php:
```diff
@@ -0,0 +1,31 @@
+use Ray\Di\Container;
+use Ray\Di\DependencyInterface;
+
+final class NullDependendy implements DependencyInterface
```
Removed.

koriym committed:
fixup! Refactor tests to use `CompiledInjector` and `Compiler`.

koriym commented on tests/script/null_object.php:
```diff
@@ -4,14 +4,13 @@
-        return new FakeNullObjectModule();
-    }
-);
+$sciptDir = dirname(__DIR__) . '/tmp/null_object';
```
Fixed!

koriym commented on tests/CompiledInjectorExtendedScriptInjectorTest.php:
```diff
@@ -0,0 +1,323 @@
+        $this->assertInstanceOf(FakeCar::class, $car);
+    }
+
+    public function testCompileOnDemandAopSerialize(): void
+    {
+        $tmpDir = $this->getTmpDir(__FUNCTION__);
+        $scriptDir = $tmpDir;
+        (new Compiler())->compile($scriptDir, new FakeAopModule());
+        $injector = new CompiledInjector($scriptDir);
+        $aop = $injector->getInstance(FakeAopInterface::class);
+        $result = $aop->returnSame(1);
+        $this->assertSame(2, $result);
+    }
```
It's OK. Each test made each directories, and those directories will erased at once.

koriym committed:
fixup! Rename LazyModule.php and mark it as deprecated

koriym commented on tests/Fake/FakeLazyModule.php:
```diff
@@ -7,15 +7,14 @@
-        $module->install(new FakeMultiBindingsModule());
-
-        return $module;
+        $this->install(new FakeCarModule());
+        $this->install(new FakeLoggerModule());
+        $this->install(new FakeToBindSingletonModule());
+        $this->install(new FakeMultiBindingsModule());
+        $this->install(new FakeCarModule());
```
Fixed!

koriym commented on tests/ContextInjectorTest.php:
```diff
@@ -25,10 +25,10 @@ public function testGetRayInjector(): InjectorInterface
 
-    public function testGetCompileInjector(): void
+    public function testGetCompiledInjector(): void
     {
         $injector = ContextInjector::getInstance(new FakeProdContext(__DIR__ . '/tmp/base'));
-        $this->assertInstanceOf(CompileInjector::class, $injector);
+        $this->assertInstanceOf(CompiledInjector::class, $injector);
```
Fixed!

koriym replied:
Go ahead!

koriym replied:
Done with https://github.com/ray-di/Ray.Compiler/pull/115#:~:text=%E2%80%A6-,9c6ec91,-Merge%20info

koriym replied:
Please!

koriym replied:
Thanks. Fixed with https://github.com/ray-di/Ray.Compiler/pull/115#:~:text=%E2%80%A6-,5df3f09,-Loading

koriym commented on tests/CompiledInjectorTest.php:
```diff
@@ -66,8 +67,10 @@
         deleteFiles(__DIR__ . '/tmp');
-        /** @var CompileInjector $injector */
-        $injector = unserialize(serialize(new CompileInjector(__DIR__ . '/tmp', new FakeLazyModule())));
+        $scriptDir = __DIR__ . '/tmp';
+        (new Compiler())->compile($scriptDir, new FakeLazyModule());
+        $injector = new CompiledInjector($scriptDir);
+        $injector = unserialize(serialize($injector));
```
Fixed!

koriym committed:
Update src/CompiledInjector.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

koriym commented on composer.json:
```diff
@@ -12,20 +12,20 @@
-        "phpunit/phpunit": "^8.5.24 || ^9.5",
-        "bamarni/composer-bin-plugin": "^1.4"
+        "bamarni/composer-bin-plugin": "^1.4",
+        "maglnet/composer-require-checker": "^4.14",
```
Removed maglnet/composer-require-checker. Fixed!

koriym commented on composer-require-checker.json:
```diff
@@ -4,19 +4,14 @@
-    "Ray\\Di\\ProviderInterface", "Ray\\Di\\ProviderSetModule",
-    "Ray\\Di\\SetContextInterface"
-
+    "PhpParser",
+    "PhpParser\\BuilderFactory", "PhpParser\\Node", "PhpParser\\NodeAbstract", "PhpParser\\Node\\Arg",
+    "PhpParser\\Node\\Expr", "PhpParser\\Node\\Expr\\ArrayItem", "PhpParser\\Node\\Expr\\Array_",
+    "PhpParser\\Node\\Expr\\Array_", "PhpParser\\Node\\Expr\\Assign", "PhpParser\\Node\\Expr\\ConstFetch",
+    "PhpParser\\Node\\Expr\\FuncCall", "PhpParser\\Node\\Expr\\MethodCall", "PhpParser\\Node\\Expr\\New_",
+    "PhpParser\\Node\\Expr\\PropertyFetch", "PhpParser\\Node\\Expr\\Variable", "PhpParser\\Node\\Name",
+    "PhpParser\\Node\\FullyQualified", "PhpParser\\Node\\Scalar\\DNumber", "PhpParser\\Node\\Scalar\\LNumber",
+    "PhpParser\\Node\\Scalar\\String_", "PhpParser\\Node\\Stmt\\Return_", "PhpParser\\ParserFactory",
+    "PhpParser\\PrettyPrinter\\Standard", "PhpParser\\Node\\Name\\FullyQualified"
```
True. But composer-require-checker scan that deprecated directory. Any better way?

koriym replied:
Hey, No comment in JSON!

koriym replied:
Necessary?

koriym replied:
Then, help me please.

koriym replied:
Not enough. Should be...

```json
{
  "symbol-whitelist": [
    "null", "true", "false",
    "static", "self", "parent",
    "array", "string", "int", "float", "bool", "iterable", "callable", "void", "object",
    "Attribute", "ReflectionAttribute",
    "PhpParser",
    "PhpParser\\BuilderFactory",
    "PhpParser\\Node",
    "PhpParser\\NodeAbstract",
    "PhpParser\\Node\\Arg",
    "PhpParser\\Node\\Expr",
    "PhpParser\\Node\\Expr\\ArrayItem",
    "PhpParser\\Node\\Expr\\Array_",
    "PhpParser\\Node\\Expr\\Assign",
    "PhpParser\\Node\\Expr\\ConstFetch",
    "PhpParser\\Node\\Expr\\FuncCall",
    "PhpParser\\Node\\Expr\\MethodCall",
    "PhpParser\\Node\\Expr\\New_",
    "PhpParser\\Node\\Expr\\PropertyFetch",
    "PhpParser\\Node\\Expr\\Variable",
    "PhpParser\\Node\\Name",
    "PhpParser\\Node\\FullyQualified",
    "PhpParser\\Node\\Scalar\\DNumber",
    "PhpParser\\Node\\Scalar\\LNumber",
    "PhpParser\\Node\\Scalar\\String_",
    "PhpParser\\Node\\Stmt\\Return_",
    "PhpParser\\ParserFactory",
    "PhpParser\\PrettyPrinter\\Standard",
    "PhpParser\\Node\\Name\\FullyQualified"
  ]
}
```

koriym committed:
Refactor test cleanup with variable for script directory

Replace repeated directory paths with a variable to improve readability and maintainability. This ensures consistent usage of the script directory across test methods.

koriym committed:
Add a missing template annotation in CompiledInjector.php

The added `@template T` annotation improves type safety and clarity. This change ensures better compatibility with static analysis tools like PHPStan and enhances code maintainability.

koriym committed:
Refactor InjectorFactory to simplify injector creation logic

Replaced redundant methods and updated method calls for clarity and consistency. Consolidated `getScriptInjector` and `getCompileInjector` into `getCompiledInjector` to streamline injection process handling. This improves code maintainability and reduces duplication.

koriym commented on tests/CompiledInjectorTest.php:
```diff
@@ -7,29 +7,30 @@
     }
 
-    /** @var CompileInjector $injector */
+    /** @var CompiledInjector $injector */
```
Done.

koriym committed:
Rename FakeLazyModule to FakeModule and update references

Replaced all references to `FakeLazyModule` with `FakeModule` for consistency and clarity. This includes renaming the file, class, and test cases, ensuring proper alignment across the codebase.

koriym committed:
Refactor test temp directories to use test method names.

Updated temporary script directories in tests to include the respective test method names. This change ensures better isolation between tests and avoids potential conflicts when running tests simultaneously.

koriym commented on tests/CompiledInjectorExtendedScriptInjectorTest.php:
```diff
@@ -0,0 +1,323 @@
+        $this->assertInstanceOf(FakeRobot::class, $robot);
+    }
+
+    /** @return non-empty-string */
+    private function getTmpDir(string $subDirectory): string
+    {
+        $tmpDir = __DIR__ . '/tmp/' . $subDirectory;
+        @mkdir($tmpDir);
+
+        return $tmpDir;
+    }
```
Not necessary. Each directory has a mechanism to keep cleaed.

koriym committed:
Update class reference in CompiledInjector docblock

Replaced the outdated reference to `Ray\Compiler\CompileInjector` with the correct `Ray\Compiler\Compiler`. This ensures the documentation aligns with the current implementation.

koriym committed:
fixup! Remove obsolete testCompileFailureLog function

koriym commented on src/CompiledInjector.php:
```diff
@@ -0,0 +1,122 @@
+            return $this->singletons[$dependencyIndex];
+        }
+
+        $scriptFile = sprintf('%s/%s.php', $this->scriptDir, str_replace('\\', '_', $dependencyIndex));
+        if (! file_exists($scriptFile)) {
+            throw new Unbound($dependencyIndex); // Binding not found
+        }
```
Current prevention is enough.

koriym replied:
> $isProd = (bool) $rayInjector->getInstance('', Compile::class);

Done!

koriym commented:
@coderrabitai Approved?

NaokiTsuchiya commented on README.md:
```diff
@@ -2,67 +2,173 @@
+use Ray\Compiler\Compiler;
+
+$compiler = new Compiler();
+$compiler->compile($module, $scriptDir);
```
以下のメソッドシグネチャみたいです！

```
public function compile(string $scriptDir, AbstractModule $module): Scripts
```

```suggestion
$compiler->compile($scriptDir, $module);
```

NaokiTsuchiya commented on README.md:
```diff
@@ -2,67 +2,173 @@
+use Ray\Compiler\Compiler;
+
+$compiler = new Compiler();
+$compiler->compile(new DevModule(), __DIR__ . '/tmp/di');
```
```suggestion
$compiler->compile(__DIR__ . '/tmp/di', new DevModule());
```

koriym replied:
あー！これ戻したのが悪かったですね。Injectorと同じく`($module, $dir)にしましょうか。

koriym replied:
元に戻しました＞＜

koriym committed:
Refactor compile method to swap parameter order

Changed the order of parameters for the `Compiler::compile` method, placing the module before the script directory. Updated all relevant calls, tests, and documentation to reflect this change for consistency and clarity.

koriym committed:
Refactor script path handling to use relative paths

Replaced absolute script directory paths with relative paths for better portability, particularly in containerized environments. Introduced a utility method to compute relative paths and updated all relevant references in the code. This improves flexibility and ensures compatibility across different deployment setups.

koriym commented:
@NaokiTsuchiya https://github.com/ray-di/Ray.Compiler/pull/115/commits/8a70d8d8c9b180723e5dda3f79f20fc541862a0c scriptDirを相対パスにしたのでこれで（シリアライズしたとしても）コンテナ環境でも大丈夫だと思います！

...と思ったのですがやはりdockerでパスを揃える運用の方がストレートだと思いrevertしました。

koriym committed:
Add documentation for ScriptDirNotReadable error

This new document explains the causes and solutions for the ScriptDirNotReadable error in both local and Docker environments. It includes examples, troubleshooting steps, and a quick checklist to help users resolve the issue efficiently.

koriym committed:
Add detailed error documentation and improve exception message

Added a new documentation file for the "Unbound Error," explaining causes and solutions in-depth. Enhanced the `ScriptDirNotReadable` exception message to provide a helpful URL for troubleshooting.

koriym committed:
Refactor README to improve clarity and streamline instructions

Reorganized and simplified sections to make the README easier to follow. Removed redundant content and emphasized key usage instructions for production and development. Enhanced Docker and version control guidance, ensuring best practices are clearly communicated.

koriym committed:
Revert "Refactor script path handling to use relative paths"

This reverts commit 8a70d8d8c9b180723e5dda3f79f20fc541862a0c.

koriym committed:
Implement Countable interface in Scripts class

Added the Countable interface to the Scripts class along with a `count` method, providing the ability to count the number of scripts. This enhances the usability of the class for operations requiring a countable structure.

koriym committed:
Add compile script for dependency injection container

Introduced a new PHP script to precompile dependency injection configurations using Ray.Di. This improves performance by generating container files ahead of time and storing them in the `di` directory.

koriym committed:
Add compile script for dependency injection container

Introduced a new PHP script to precompile dependency injection configurations using Ray.Di. This improves performance by generating container files ahead of time and storing them in the `di` directory.

koriym committed:
Add example for creating a compile script in README

This update enhances the README by providing an example of how to create a compile script using the Ray Compiler. It includes a PHP code snippet and improves clarity for users integrating the compiler.

koriym committed:
Add unit tests for Scripts class

Introduced `ScriptsTest` to validate the functionality of the `Scripts` class, including methods like `add`, `count`, and `save`. Additionally, fixed a minor syntax error in the README example.

koriym commented on bin/compile.php:
```diff
@@ -0,0 +1,13 @@
+
+require dirname(__DIR__) . '/vendor/autoload.php';
+
+$scripts = (new Compiler())->compile(
+    new FakeCarModule(),
+    __DIR__ . '/di'
+);
```
Too much!

koriym committed:
Cast file operations to ensure type safety

Explicitly cast file_get_contents and glob outputs to string and array, respectively, to guarantee type consistency. Added a PHPStan ignore comment to suppress type-checking warnings for the cast on glob.

koriym commented on README.md:
```diff
@@ -1,68 +1,81 @@
+    "scripts": {
+        "post-install-cmd": ["php bin/compile.php"]
+    }
+}
```
Fixed!

koriym committed:
fixup! Add example for creating a compile script in README

koriym commented on bin/compile.php:
```diff
@@ -0,0 +1,13 @@
+<?php
+
+use Ray\Compiler\Compiler;
+use Ray\Compiler\FakeCarModule;
+
+require dirname(__DIR__) . '/vendor/autoload.php';
```
Done!

koriym committed:
Fix punctuation in README.md for list consistency

Added missing periods to ensure consistent punctuation in the list of components. This improves readability and maintains professional documentation standards.

koriym committed:
Update README with Dockerfile improvements and .gitignore

Clarified Dockerfile instructions by adding working directory setup, DI compilation step, and precise file copying between stages. Updated `.gitignore` snippet for correct syntax highlighting.

koriym committed:
Update temp directory path for ScriptsTest

Replaced `sys_get_temp_dir` with a project-relative path to ensure consistent and predictable directory usage during tests. Updated the class docblock to provide a clearer explanation of its purpose.

koriym commented on src/Scripts.php:
```diff
@@ -0,0 +1,41 @@
+        $this->scripts[$index] = $script;
+    }
+
+    public function save(string $scriptDir): void
+    {
+        $template = <<<'EOL'
+<?php
+%s
+EOL;
+        $filePutContents = new FilePutContents();
+        foreach ($this->scripts as $index => $script) {
+            $file = sprintf('%s/%s.php', $scriptDir, str_replace('\\', '_', $index));
+            $script = sprintf($template, $script);
+            $filePutContents($file, $script);
+        }
+    }
```
Done!

koriym committed:
fixup! Update README with Dockerfile improvements and .gitignore

koriym committed:
Refactor README with clearer examples and error handling.

Updated code examples to improve clarity, including detailed comments and enhanced error handling for the compile script. Enhanced Dockerfile setup with better dependency management, non-root user support, and streamlined file copying.

koriym committed:
Update PHP version to 8.4 in GitHub workflows

Updated the PHP version from 8.3 to 8.4 in the static-analysis and coding-standards workflows. This change ensures compatibility with the latest PHP features and improvements.

koriym committed:
Update dev dependencies and add new packages for Amp v3

Updated dependencies to support PHP 8.1 and newer versions of Amp packages, including `phpmetrics`, `psalm`, and `amphp/byte-stream`. Introduced additional Amp libraries like `amphp/cache`, `amphp/dns`, and others to enhance async capabilities. Adjusted project configurations accordingly.

koriym committed:
Improve Dockerfile and update README with best practices.

Refactored the Dockerfile for better build optimization, added a non-root user setup, and provided a recommended `.dockerignore` example. Updated the README to enhance clarity and emphasize best practices for handling environment-specific artifacts.
